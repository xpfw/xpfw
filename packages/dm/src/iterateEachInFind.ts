import { cloneDeep, get, isArray, isObject } from "lodash"

export interface IEachInFindOptions {
  pageSize?: number
}
const iterateEachInFind =
async (collection: string | string[],
       queryObj: any, findMethod: (c: string, q: any) => Promise<any>, iterator: (obj: any) => Promise<any>,
       options?: IEachInFindOptions) => {
  console.log(`entering with collection", collection`)
  if (!isObject(queryObj)) {
    queryObj = {}
  }
  if (collection != null && !Array.isArray(collection)) {
    collection = [collection]
  }
  const pageSize: any = get(options, "pageSize", 5)
  console.log("Trying to iterate", collection)
  for (const col of collection) {
    let gotMorePages = true
    let currentPage = 0
    let res
    while (gotMorePages) {
      const queryWith = cloneDeep(queryObj)
      queryWith.$limit = pageSize
      queryWith.$skip = pageSize * currentPage
      const findRes = await findMethod(col, queryWith)
      if (findRes != null) {
        if (isArray(findRes.data)) {
          for (const e of findRes.data) {
            e.fromCollection = col
            res = await iterator(e)
          }
        }
        currentPage++
        if (currentPage >= findRes.total / pageSize) {
          gotMorePages = false
        }
      } else {
        gotMorePages = false
      }
    }
    return res
  }
}

export default iterateEachInFind
