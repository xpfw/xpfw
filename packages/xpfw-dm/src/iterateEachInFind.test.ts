import iterateEachInFind from "./iterateEachInFind"

it("Expect iterateEachInFind behaves as expected", async () => {
  const findMethod = (c: string, q: any) => {
    if (q.$skip === 0) {
      return Promise.resolve({total: 10, data: [{a: "b"}, {q: "b"}, {h: "b"}, {e: "b"}, {a: "c"}]})
    }
    return Promise.resolve({total: 15, data: [{jh: "we"}, {ju: "3"}, {k: "4"}, {l: "รถ"}, {r: "e"}]})
  }
  let iteratedItems: any[] = []
  let i = 0
  await iterateEachInFind("bla", {qu: "la"}, findMethod, (obj: any) => {
    iteratedItems.push({sort: i, item: obj})
    i++
    return Promise.resolve()
  }, {pageSize: 5})
  expect(iteratedItems).toMatchSnapshot("result of iterate")
  iteratedItems = []
  await iterateEachInFind("bla", undefined, findMethod, (obj: any) => {
    iteratedItems.push({sort: i, item: obj})
    i++
    return Promise.resolve()
  }, {pageSize: 5})
  expect(iteratedItems).toMatchSnapshot("result of no query iteration")
  iteratedItems = []
  await iterateEachInFind("bla", undefined, () => Promise.resolve(null), (obj: any) => {
    iteratedItems.push({sort: i, item: obj})
    i++
    return Promise.resolve()
  }, {pageSize: 5})
  expect(iteratedItems).toMatchSnapshot("result of no faulty find iteration")
})
