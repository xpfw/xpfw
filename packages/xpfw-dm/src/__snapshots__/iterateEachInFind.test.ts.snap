// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect iterateEachInFind behaves as expected: result of iterate 1`] = `
Array [
  Object {
    "item": Object {
      "a": "b",
      "fromCollection": "bla",
    },
    "sort": 0,
  },
  Object {
    "item": Object {
      "fromCollection": "bla",
      "q": "b",
    },
    "sort": 1,
  },
  Object {
    "item": Object {
      "fromCollection": "bla",
      "h": "b",
    },
    "sort": 2,
  },
  Object {
    "item": Object {
      "e": "b",
      "fromCollection": "bla",
    },
    "sort": 3,
  },
  Object {
    "item": Object {
      "a": "c",
      "fromCollection": "bla",
    },
    "sort": 4,
  },
  Object {
    "item": Object {
      "fromCollection": "bla",
      "jh": "we",
    },
    "sort": 5,
  },
  Object {
    "item": Object {
      "fromCollection": "bla",
      "ju": "3",
    },
    "sort": 6,
  },
  Object {
    "item": Object {
      "fromCollection": "bla",
      "k": "4",
    },
    "sort": 7,
  },
  Object {
    "item": Object {
      "fromCollection": "bla",
      "l": "รถ",
    },
    "sort": 8,
  },
  Object {
    "item": Object {
      "fromCollection": "bla",
      "r": "e",
    },
    "sort": 9,
  },
  Object {
    "item": Object {
      "fromCollection": "bla",
      "jh": "we",
    },
    "sort": 10,
  },
  Object {
    "item": Object {
      "fromCollection": "bla",
      "ju": "3",
    },
    "sort": 11,
  },
  Object {
    "item": Object {
      "fromCollection": "bla",
      "k": "4",
    },
    "sort": 12,
  },
  Object {
    "item": Object {
      "fromCollection": "bla",
      "l": "รถ",
    },
    "sort": 13,
  },
  Object {
    "item": Object {
      "fromCollection": "bla",
      "r": "e",
    },
    "sort": 14,
  },
]
`;

exports[`Expect iterateEachInFind behaves as expected: result of no faulty find iteration 1`] = `Array []`;

exports[`Expect iterateEachInFind behaves as expected: result of no query iteration 1`] = `
Array [
  Object {
    "item": Object {
      "a": "b",
      "fromCollection": "bla",
    },
    "sort": 15,
  },
  Object {
    "item": Object {
      "fromCollection": "bla",
      "q": "b",
    },
    "sort": 16,
  },
  Object {
    "item": Object {
      "fromCollection": "bla",
      "h": "b",
    },
    "sort": 17,
  },
  Object {
    "item": Object {
      "e": "b",
      "fromCollection": "bla",
    },
    "sort": 18,
  },
  Object {
    "item": Object {
      "a": "c",
      "fromCollection": "bla",
    },
    "sort": 19,
  },
  Object {
    "item": Object {
      "fromCollection": "bla",
      "jh": "we",
    },
    "sort": 20,
  },
  Object {
    "item": Object {
      "fromCollection": "bla",
      "ju": "3",
    },
    "sort": 21,
  },
  Object {
    "item": Object {
      "fromCollection": "bla",
      "k": "4",
    },
    "sort": 22,
  },
  Object {
    "item": Object {
      "fromCollection": "bla",
      "l": "รถ",
    },
    "sort": 23,
  },
  Object {
    "item": Object {
      "fromCollection": "bla",
      "r": "e",
    },
    "sort": 24,
  },
  Object {
    "item": Object {
      "fromCollection": "bla",
      "jh": "we",
    },
    "sort": 25,
  },
  Object {
    "item": Object {
      "fromCollection": "bla",
      "ju": "3",
    },
    "sort": 26,
  },
  Object {
    "item": Object {
      "fromCollection": "bla",
      "k": "4",
    },
    "sort": 27,
  },
  Object {
    "item": Object {
      "fromCollection": "bla",
      "l": "รถ",
    },
    "sort": 28,
  },
  Object {
    "item": Object {
      "fromCollection": "bla",
      "r": "e",
    },
    "sort": 29,
  },
]
`;
